src/main.cpp

```diff
@@ -6355,7 +6355,26 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
 
         bool fCanDirectFetch = CanDirectFetch(chainparams.GetConsensus());
         CNodeState *nodestate = State(pfrom->GetId());
-        nodestate->fFirstHeadersReceived = true;
+
+        // During the initial peer handshake we must receive the initial headers which should be greater
+        // than or equal to our block height at the time of requesting GETHEADERS. This is because the peer has
+        // advertised a height >= to our own. Furthermore, because the headers max returned is as much as 2000 this
+        // could not be a mainnet re-org.
+        if (!nodestate->fFirstHeadersReceived)
+        {
+            // We want to make sure that the peer doesn't just send us any old valid header. The last header block
+            // hash they send us should now be equal to our current pindexLast block hash.
+            if (pindexLast && pindexLast->IsValid(BLOCK_VALID_TREE) &&
+                chainActive.Tip()->nChainWork <= pindexLast->nChainWork)
+            {
```

Comment:
I would feel better if the test were for the actual claim the peer made - that they have blocks to (at least) a given height.  For example, if there is a temporary split in the chain from a natural block race, it is possible they are on one side of the fork and we are on the other.  If this happened around the 2016-block difficulty reset, one side can have lower difficulty than the other (all you need is a different block timestamp) and this code would then be penalizing the peer for being on the other side - it may not be aware of our higher-chainwork side.  This is a minor point perhaps but worth keeping in mind.